### ISSUE TITLE
waitanyinvoice incorrectly coded (definitely)

### ISSUE THREAD

**Issue by ZmnSCPxj** [ElementsProject/lightning#442]
A quick review of the `waitanyinvoice` and `waitinvoice` commands suggests they are not correctly coded.  I reviewed these due to: https://github.com/ElementsProject/lightning/issues/439

I copy my analyses here:

Looking through the `json_waitanyinvoice` and `invoice.c` in general, it seems the behavior is something like this:

1.  `waitanyinvoice` will skip over any unpaid invoices.
2.  New invoices are put at the head of the list of invoices.
3.  Payment of invoices does not change the ordering of the list.
4. Given an argument, `waitanyinvoice` will scan the list starting from the given label,

The intent is that `waitanyinvoice` will be called repeatedly, with the first call given no argument, and succeeding calls being given as argument the `label` property of the previous return value of `waitanyinvoice`.

So it seems to me (unsure) that it would be possible for `waitanyinvoice` to report invoices multiple times.  For example:

1.  2 invoices are created, A and B.  Because new invoices are created at the head of the invoice list, this means the list is [B, A].
2.  `waitanyinvoice` is called once without an argument.
3.  The code scans over the list.  Since the none of the invoices are paid, the call blocks until any invoice is paid.
4.  Invoice A is paid.
5.  This triggers the `tell_waiter` function.  It then reports invoice A.
6.  `waitanyinvoice` is called with argument A (the previous return value).
7.  A is the last invoice in the list and we reach end-of-list.  So the call blocks until any invoice is paid.
8.  Invoice B is paid.
9.  This triggers `tell_waiter`, it reports invoice B.
10.  `waitanyinvoice` is called with argument B (the previous return value).
11.  The code starts with the invoice B.  B is the first item in the list [B, A].  So it looks starting at the next item.
12.  The next item after invoice B is invoice A, which is a paid invoice.  We dutifully report this invoice back to the caller without blocking.
13.  Oops, invoice A is credited twice by BitPay losing trillions of BTC!!!!!!!

@cdecker @rustyrussell can you confirm my analysis is correct?

If my analysis is correct, I think we should have a separate list of paid and unpaid invoices.

When an invoice is paid, it is removed from the unpaid invoice list, and added to the *end* of the paid invoice list.

Then `waitanyinvoice` should scan the paid invoice list only.

Also, `waitinvoice` also uses the same mechanism as `waitanyinvoice` (I think), so if `waitinvoice` is invoked, the invoice specified is unpaid, and then a *different* invoice is paid, it seems to my analysis that `waitinvoice` will return the other invoice invoice, not the one it was told to scan.

----

Actually if my analysis is correct, it may be possible for waitanyinvoice to block while an invoice is paid that has not been encountered by the caller to waitanyinvoice under a race condition.

1.  Make invoices A and B, creating the internal list [B, A]
2.  Pay A.
3.  waitanyinvoice with no argument, returning A.
4.  BtcPay processes etc etc
5.  Pay B.
6.  BtcPay finishes processing.
7.  waitanyinvoice with argument A.  Since A comes after B in the internal list, c-lightning assumes B is already reported and blocks.
8.  Payer of B wonders why BtcPay is not working and loses trillions of BTC due to lost business opportunity.

But in any case, waitanyinvoice "should" be something c-lightning provides, it's just that it might be buggy for now. If I can confirm my analysis is correct, I shall open a new issue.

----

I will have to actually run the code and see if my analysis is correct, or maybe @rustyrussell and @cdecker can confirm/disconfirm my analysis.

**Comment by ZmnSCPxj** [comment]
Reviewing again, I made some mistake in my analysis:

1.  Adding an invoice adds to the tail of the invoice list, not head: https://github.com/ElementsProject/lightning/blob/master/lightningd/invoice.c#L212
2.  This means, if we create invoices A and B in that order, the internal invoice list is [A, B], not [B,A]

However the rest of the analysis still seems to hold:

1.  Create invoices [A, B] in that order.
2.  Pay B first.
3.  `waitanyinvoice` -> returns B
4.  Pay A.
5. `waitanyinvoice` -> unexpected stall, expected return A

Again, this seems to be, because `resolve_invoice` does not rearrange the list --- at minimum, it should detach the resolved invoice from the list, then put it again to the end of the list (and design-wise, I think I would prefer a separate list for paid invoices).

Even when we switch to using the database instead of internal invoice list, we should arrange `waitanyinvoice` to iterate in the order of payment rather than order of invoice creation.

For now, I notice there is an existing `test_waitanyinvoice`.  It pays in the same order that invoices are created, but I think this will fail if invoices are paid in the opposite order of creation.  I shall create a modified version of this  test (probably separate, in order to start with an empty internal invoice list), to pay in the opposite order as creation, and see if `waitanyinvoice` fails.

---

Note also, there seems to be some problem with `waitinvoice` also, since it uses the same system as `waitanyinvoice`, I will also need to study it further.

**Comment by ZmnSCPxj** [comment]
Confirmed: https://travis-ci.org/ZmnSCPxj/lightning/jobs/316409078

    def test_waitanyinvoice_reversed(self):
        """Test waiting for invoices, where they are paid in reverse order
        to when they are created.
        """
        # Setup
        l1, l2 = self.connect()
        self.fund_channel(l1, l2, 10**6)
        l1.bitcoin.generate_block(6)
        l1.daemon.wait_for_log('Received node_announcement for node {}'.format(l2.info['id']))

        # Create invoices
        inv1 = l2.rpc.invoice(1000, 'inv1', 'inv1')
        inv2 = l2.rpc.invoice(1000, 'inv2', 'inv2')

        # Pay inv2, wait, pay inv1, wait
        # Pay inv2
        l1.rpc.pay(inv2['bolt11'])
        # Wait - should not block, should return inv2
        r = self.executor.submit(l2.rpc.waitanyinvoice).result(timeout=5)
        assert r['label'] == 'inv2'
        # Pay inv1
        l1.rpc.pay(inv1['bolt11'])
        # Wait inv2 - should not blok, should return inv1
        r = self.executor.submit(l2.rpc.waitanyinvoice, 'inv2').result(timeout=5)
        assert r['label'] == 'inv1'

Will now also check `waitinvoice` behavior.  If `waitinvoice` is also buggy (as my analysis) then I will split this issue into two.

**Comment by ZmnSCPxj** [comment]
Yes, `waitinvoice` is also buggy: https://travis-ci.org/ZmnSCPxj/lightning/jobs/316415056

    def test_waitinvoice(self):
        """Test waiting for one invoice will not return if another invoice
        is paid.
        """
        # Setup
        l1, l2 = self.connect()
        self.fund_channel(l1, l2, 10**6)
        l1.bitcoin.generate_block(6)
        l1.daemon.wait_for_log('Received node_announcement for node {}'.format(l2.info['id']))

        # Create invoices
        inv1 = l2.rpc.invoice(1000, 'inv1', 'inv1')
        inv2 = l2.rpc.invoice(1000, 'inv2', 'inv2')

        # Start waiting on invoice 1, should block
        f = self.executor.submit(l2.rpc.waitinvoice, 'inv1')
        time.sleep(1)
        assert not f.done()
        # Pay invoice 2
        l1.rpc.pay(inv2['bolt11'])
        # Waiter should stil be blocked
        time.sleep(1)
        assert not f.done()
        # Pay invoice 1
        l1.rpc.pay(inv1['bolt11'])
        # Waiter should now finish
        r = f.result(timeout=5)
        assert r['label'] == 'inv1'

I will now split off the `waitinvoice` bug into a separate issue, and, create pull requests starting with these failing tests (and which should terminate in an actual solution)

**Comment by ZmnSCPxj** [comment]
The simplest solution is this:

1.  In `resolve_invoice`, remove the invoice from the invoices list, then put it to the end of the invoices list.

However, the following issue is not solved by this:

1.  Make invoice A and B.
2.  Pay invoice A.
3.  Client calls `waitanyinvoice` without argument -> don't block, return A immediately.
4.  Crash `lightningd` and restart it.
5.  Pay invoice B.
6.  Client calls `waitanyinvoice` with argument A -> shouldn't block, should return B.  However, assuming the db stores it in order [A, B], `invoice_add` will add it in reverse order and store the internal order [B, A] in the in-memory list.  This makes `lightningd` assume that B has been reported even if it is not yet reported.

So a full solution should arguably require moving to using the db:

1.  Add a new table `paid_invoices` with columns `ord INTEGER NOT NULL AUTO_INCREMENT` and `invoice_id INTEGER`, with `PRIMARY KEY (ord)` and `UNIQUE (ord), UNIQUE (invoice_id)`.
2.  To migrate from older database, we should execute `INSERT INTO paid_invoices (invoice_id) SELECT id FROM invoices WHERE state=1`.
3.  `waitanyinvoice`, given no argument, will do `SELECT invoice_id FROM paid_invoices LIMIT 1 ORDER BY ord ASC` to locate the invoice ID, then extract the information from the `invoices` table given the invoice ID.  If the query returns no rows, block.
4.  `waitanyinvoice`, given an argument, should locate the invoice ID of the argument, then `SELECT ord FROM paid_invoices WHERE invoice_id=?` on the invoice ID, then `SELECT invoice_id FROM paid_invoices WHERE ord>? ORDER BY ord ASC LIMIT 1` to locate the next invoice ID, then extract the information from the `invoices` table given the invoice ID.  If the query returns no rows, block.

**Comment by ZmnSCPxj** [comment]
Database should go like this:

## Add to `dbmigrations`

    CREATE TABLE paid_invoices (
        ord INTEGER,
        invoice_id INTEGER,
        PRIMARY KEY(ord),
        FOREIGN KEY(invoice_id) REFERENCES invoices(id) ON DELETE CASCADE ON UPDATE RESTRICT
    );

    INSERT INTO paid_invoices(invoice_id)
    SELECT id
    FROM invoices
    WHERE state = 1; -- paid invoices

The foreign key constraints cause the item in `paid_invoices` to be removed if the row in `invoices` is removed, e.g. by `delinvoice` command.

## Query to execute on `resolve_invoice`

    INSERT INTO paid_invoices(invoice_id) VALUES(?);

replacing 1st `?` with `invoice->id`

## Query to execute on `waitanyinvoice` without argument

    SELECT (label, rhash, msatoshi, state)
    FROM invoices INNER JOIN paid_invoices ON id = invoice_id
    ORDER BY ord ASC
    LIMIT 1;

If query results in 0 rows, block.

## Query to execute on `waitanyinvoice` with label argument

    SELECT (label, rhash, msatoshi, state)
    FROM invoices INNER JOIN paid_invoices ON id = invoice_id
    WHERE ord > (SELECT ord FROM invoices INNER JOIN paid_invoices ON id=invoice_id WHERE label=?)
    ORDER BY ord ASC
    LIMIT 1;

replacing 1st `?` with the label argument.

If the query results in 0 rows, it is either because the label doesn't exist, or the label identifies the last paid invoice.  How to differentiate?  I do not know how to do this in a single query, but with a second query we can simply check `SELECT COUNT(*) FROM invoices INNER JOIN paid_invoices ON id=invoice_id WHERE label=?` and see if it is zero or not.  The problem with this is if in the future we make `lightningd` multithreaded or multiprocess, we will need to put multiple queries in a transaction.

**Comment by cdecker** [comment]
I think we should definitely use the `invoices` table and add a paid order field to that instead of creating a table just for that case.

**Comment by ZmnSCPxj** [comment]
Using a separate table lets us have SQLite implement an incrementing order variable for us without having to wrangle it ourselves.

Otherwise we would need `dbmigrations`:

    ALTER TABLE invoices
      ADD COLUMN paid_ord INTEGER;
    CREATE UNIQUE INDEX -- SQLite does not allow altering new column with UNIQUE constraint
      invoices_paid_ord_index ON invoices(paid_ord);
    UPDATE invoices
      SET paid_ord = id WHERE state = 1;
    CREATE TABLE next_paid_ord(next_paid_ord INTEGER);
    INSERT INTO next_paid_ord
      VALUES(COALESCE(MAX(SELECT paid_ord FROM invoices WHERE state = 1), 1));

We would need to wrangle with this persistent `next_paid_ord` variable each time we want to update an invoice to paid.  Alternative we could do `UPDATE invoices SET state=1, paid_ord = COALESCE(MAX(SELECT paid_ord FROM invoices WHERE state = 1), 1)` and drop `next_paid_ord`.

Over all, simplest solution is still to use a separate table with `INTEGER PRIMARY KEY AUTOINCREMENT`.

**Comment by cdecker** [comment]
Well, we do have a `vars` table for that kind of thing, no need to create a new table just for this case.

**Comment by ZmnSCPxj** [comment]
I see.

Here is now the proposed `dbmigrations`:

    ALTER TABLE invoices ADD pay_ord INTEGER;
    CREATE UNIQUE INDEX invoices_pay_ord_idx ON invoices(pay_ord);
    UPDATE invoices SET pay_ord = id WHERE state=1;
    INSERT OR REPLACE INTO vars(name, val)
      VALUES('next_pay_ord', COALESCE((SELECT MAX(pay_ord) FROM invoices WHERE state=1),0) + 1);

Calling `waitanyinvoice` without argument:

    SELECT label FROM invoices WHERE pay_ord NOT NULL
      ORDER BY pay_ord ASC LIMIT 1;

Calling `waitanyinvoice` with an argument:

    SELECT label FROM invoices
      WHERE pay_ord NOT NULL
        AND pay_ord > (SELECT pay_ord FROM invoices WHERE label=?)
      ORDER BY pay_ord ASC LIMIT 1;

When doing `wallet_invoice_save`, if it is paid, we should check if it is being transitioned from unpaid->paid, then:

1.  Get `next_pay_ord` variable.
2.  Increment `next_pay_ord` variable. (protect against sudden crashes)
3.  Update pay_ord with the next pay_ord and update state to 1.

(all above operations would be implied in a single query if we had a separate `paid_invoices` table, and we would not have to check transition from unpaid->paid, we can just use `INSERT OR IGNORE` into the `paid_invoices` table as long as the `invoice_id` is unique)

**Comment by ZmnSCPxj** [comment]
Sorry for the long turnaround for this issue.  I am a little stuck on how to structure the code; I notice, all database operations are in `wallet/wallet.c` and there are currently none, in `lightningd/invoice.c`.

Roughly, I am making two commits:

1.  Modify database to include `pay_ord`, and update the column appropriately during migration and `wallet_invoice_save`.
2.  Modify invoice system to use this column for `waitanyinvoice`.

The latter one is making it a little difficult:

1. `wallet/wallet.c` currently has no access to the `invoices` list directly, so cannot get the in-memory invoice object.
2.  `lightningd/invoice.c` currently does not access database directly, so cannot query the state of the database (and would require a linear search of the invoice list in addition, if we want to reuse `tell_waiter`.

I am currently thinking, that the `wallet/wallet.c` will add a `wallet_invoice_waitany` function that returns the key fields `label`, `msatoshi`, and `rhash` (`complete` is known to be true).

However this requires greater change if we then switch to using `pay_ord` directly in `waitanyinvoice`, as @rustyrussell proposes....

------

Another idea, is to drop the list of struct invoice entirely.  Our primary queries to this list involve:

1.  Searching an unpaid invoice by `rhash` (for incoming payments).
2.  Searching an invoice by `label` (for `waitinvoice`, or `waitanyinvoice` if it still uses the `label` interface instead of `pay_ord`).
3.  Searching a paid invoice by `pay_ord` (for `waitanyinvoice`).

All of these queries are not done very well by a linked list.

One thing we can do, is to create a SQLite temporary table, which contains pointers to the invoice objects, and a foreign key that is the `invoices(id)` column:

    CREATE TEMPORARY TABLE invoice_ptrs
      ( id PRIMARY KEY INTEGER REFERENCES invoices(id)
      , ptr INTEGER
      );

The `ptr` column is the `(s64)(struct invoice*)` of the pointer to the allocated invoice object.  Searches using `rhash`, `label`, and `pay_ord` keys simply require a search in the `invoices` for the `id` followed by a search of the `invoice_ptrs` table to get the pointer.  For example, to query the unpaid invoice by `rhash`:

    SELECT ptr
      FROM invoice_ptr INNER JOIN invoices ON invoice_ptr(id) = invoices(id)
     WHERE state=0
       AND rhash=?;

Construction of a `struct invoices` by `invoices_init` would create the table, and register a destructor for it that deletes the table.  To support the theoretical possibility of multiple `struct invoices` objects, we can append the numeric value of the pointer to `struct invoices` to the name of `invoice_ptrs` table.

**Comment by cdecker** [comment]
I like the second approach much better, because in the long run we will be removing the in-memory invoice list altogether. The second part of that approach is a bit confusing to me. This is all single-threaded so there is actually no need for all of this complexity. I'm thinking about having a `wallet_invoice_next` that given a high water mark for the last seen invoice just returns the next higher one:

1. If there is one, great just return that one
2. If there isn't return `NULL` and register your interest in getting a notification when one comes in

In the latter case we just check, when updating an invoice and setting it as complete, whether there are any parked watchers that want to be notified, i.e., return the invoice.

**Comment by ZmnSCPxj** [comment]
I am uncertain here what you mean by "second approach".  Presumably you mean the `CREATE TEMPORARY TABLE` technique, as you refer to "removing the in-memory invoice list"? What complexity do you refer to?

**Comment by cdecker** [comment]
With second I meant the proposal below the horizontal bar, and the complexity refers to having an in-memory table with address pointers. There is no need for explicit invoice passing, just mark the next invoice as completed and notify the waiting JSON-RPC connection that there is a new invoice it might want to deliver.

**Comment by ZmnSCPxj** [comment]
The second proposal is an overhaul of the entire invoices system, not just `waitanyinvoice`. The functionality below is currently implemented by linear scan of the list:

1.  `find_unpaid`, which returns the invoice given an `rhash` if it is not expired and not yet paid.
2.  `find_invoice_by_label`, which returns the invoice given a `label`.

By using an in-memory table we can keep the same functions above (which return `struct invoice *`) but replace their internals with an in-memory database table lookup.  There would also have to be significantly more encapsulation, i.e. return `const` pointers since the caller should not directly manipulate the structures, and change `state` with a function rather than modifying it directly and then calling `wallet_invoice_save` to update the database (especially since modifying `state` should also modify the `pay_index` field if it is transitioning from `UNPAID` to `PAID` - doing this in `wallet_invoice_save` seems misnamed then, since the name implies it only saves (reads the memory and writes to disk) but it also must now update the in-memory object to set the `pay_index` field).

As an overhaul however it is admittedly drastic.

For now I am just going to implement a "simple" `wallet_invoice_waitany`.

**Comment by cdecker** [comment]
Sounds good to me, sorry for the nitpicking, I am just a bit confused about the necessity of the `invoice_ptr` table :-) I was just thinking of instead keeping the jcon hanging and pinging it when invoices in the DB change.
